packages/collapsible-side-nav/src/SideNavGroup.tsx

```diff
@@ -0,0 +1,176 @@
+import React, { useCallback, useContext, useMemo, useState } from 'react';
+import PropTypes from 'prop-types';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { HTMLElementProps, IdAllocator, OneOf } from '@leafygreen-ui/lib';
+import { uiColors } from '@leafygreen-ui/palette';
+import { SideNavContext, SideNavGroupContext } from './contexts';
+import { GlyphElement, transitionDurationMilliseconds } from './utils';
+
+const sideNavGroupStyle = css`
+  display: flex;
+  flex-direction: column;
+  margin: 15px 0;
+  border: 0 solid ${uiColors.gray.light2};
+  transition: all ${transitionDurationMilliseconds}ms ease-in-out;
+`;
+
+const sideNavGroupCollapsedStyle = css`
+  margin: 0;
+`;
+
+const sideNavGroupWithGlyphCollapsedStyle = css`
+  border-top-width: 1px;
+
+  &:last-of-type {
+    border-bottom-width: 1px;
+  }
+`;
+
+const sideNavGroupCollapsedAndCurrentStyle = css`
+  background-color: ${uiColors.green.light3};
+`;
+
+const sideNavGroupHeaderStyle = css`
+  display: flex;
+  align-items: center;
+  margin-bottom: 4px;
+  padding: 0 16px;
+  text-transform: uppercase;
+  white-space: nowrap;
+  font-size: 11px;
+  font-weight: bold;
+  line-height: 16px;
+  color: ${uiColors.green.dark2};
+`;
+
+const sideNavGroupHeaderCollapsedStyle = css`
+  margin-bottom: 0;
+`;
+
+const sideNavGroupHeaderWithGlyphCollapsedStyle = css`
+  height: 40px;
+`;
+
+const glyphStyle = css`
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  margin-right: 4px;
+`;
+
+type Props = {
+  className?: string;
+  children: React.ReactNode;
+  glyph?: GlyphElement;
+} & OneOf<{ label: string }, { label: React.ReactNode; 'aria-label': string }> &
+  Omit<HTMLElementProps<'div'>, 'aria-label' | 'ref'>;
+
+const idAllocator = IdAllocator.create('side-nav-group');
+
+const SideNavGroup = React.forwardRef<HTMLDivElement, Props>(
+  function SideNavGroup(
+    {
+      className,
+      children,
+      glyph,
+      label,
+      'aria-label': ariaLabel,
+      ...rest
+    }: Props,
+    ref,
+  ) {
+    const { collapsed, hovered, currentPath } = useContext(SideNavContext);
+
+    const id = useMemo(() => idAllocator.generate(), []);
+
+    const [containedPaths, setContainedPaths] = useState<Set<string>>(
+      () => new Set(),
+    );
+
+    const addPath = useCallback((path: string) => {
+      setContainedPaths(containedPaths => {
+        if (!containedPaths.has(path)) {
+          const updatedPaths = new Set(containedPaths);
```

Comment:
What is the benefit of making a new set each time? If we've checked that the existing set does not have a path, why can't we just add it to the containedPaths set?

Reply:
Another question: My understanding was that the purpose of using Sets was to enforce a uniqueness of paths, such that we are able to determine an active Item based on whatever path is passed to the entire Navigation. If that assumption is correct (and perhaps I'm misunderstanding) I have two questions: 

1. Right now, it seems that we're only enforcing this uniqueness within a Group, so duplicate paths can exist outside of the same group
2. Is there a world in which we'd want to support a group that has items that lead to the same page. If that's the case, does it make more sense for Items to accept an active prop rather than have the Navigation component being responsible for determining what's active?

Reply:
Another argument for perhaps supporting the active prop: In .design (for example) we have three routes that map back to the same Item being active.  I need `/component/checkbox/example` and `/component/checkbox/guideline` both to make the Checkbox Item appear active in the SideNav.

Reply:
> why can't we just add it to the containedPaths set

this would not trigger a rerender since the updated `containedPaths` would equal the previous one.

> My understanding was that the purpose of using Sets was to enforce a uniqueness of paths

The purpose of using sets here is to be able to quickly look up whether an item exists. We could use plain objects for this too, but the Set API is better-suited for this.

> we have three routes that map back to the same Item being active

I don't see how an `active` prop would add additional value. the existing logic seems to look like
```jsx
<SideNavItem active={item === activePage} ...
```
which would translate to
```jsx
<SideNav currentPath={activePage} ... >
  <SideNavItem path={item} ...

Reply:
In my case for example I'd like to do: 

```
<SideNav>
   <SideNavItem active={path.includes(componentName)}>{componentName}</SideNavItem>
</SideNav>
```

Reply:
(thank you for the other clarification points)

Reply:
I think the same can be accomplished by accepting a matching function for `currentPath`, but having something like `path.includes(componentName)` seems more unsound, e.g. `'/components/checkbox'.includes('box')`, compared to normalizing the path, e.g. `normalize('/component/checkbox/example') === '/component/checkbox'`

Reply:
Accepting a matching function would wfm:
```
<SideNav currentPath={(itemPath) => normalizePath(currentPath) === normalizePath(itemPath}

<SideNavItem path="/component/componentName/example"> 
}
```

Reply:
If you have `normalizePath` then you wouldn't need a matching function
```
<SideNav currentPath={normalizePath(currentPath)} ...
  <SideNavItem path="/component/checkbox" ...